New Feature: Response Templates

Zhao-Andy opened the pull request:
## What type of PR is this? (check all applicable)

- [x] Feature

## Description
This adds the canned response feature to be used by anyone as well as moderators. Canned responses are templates that users can save and reuse to reply to a comment.

Moderators can use canned responses made by the DEV Team to leave a comment as Sloan. These "moderator canned responses" are meant to be used to help reply in situations where posting under Sloan might be more effective than as the moderator themselves.

To do:
- [ ] changelog post
- [x] write tests 😓 
- [ ] add new site config value for field `user_moderator_id` to production
- [x] make this work for replies (currently only works for new, top-level comments)

## Related Tickets & Documents
#5000 

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
![2020-01-09 12 25 53](https://user-images.githubusercontent.com/17884966/72090082-5005b780-32db-11ea-86ee-84ccf3a25288.gif)
![Screen Shot 2020-01-09 at 11 30 16 AM](https://user-images.githubusercontent.com/17884966/72090139-6e6bb300-32db-11ea-858e-32d2bb59be7e.png)



## Added to documentation?
- [x] no documentation needed (maybe?)

rhymes commented on app/javascript/cannedResponses/cannedResponses.js:
```diff
@@ -0,0 +1,215 @@
+          ${modResponseHTML}
+          <header><h3>Personal Templates</h3></header>
+          ${personalResponseHTML}
+          <a target="_blank" rel="noopener nofollow" href="/settings/canned-responses" class="mod-respons-create-new">Create new template</a>
```
this link is broken

rhymes commented:
BTW Codeclimate is complaining because `app/javascript/cannedResponses/cannedResponses.js` contains inner functions that lengthen the length of the outer function. Since you're using javascript modules, you don't need to hide the functions inside `initCannedResponses`

Zhao-Andy commented on app/controllers/comments_controller.rb:
```diff
@@ -110,6 +110,35 @@ def create
+      Notification.send_new_comment_notifications_without_delay(@comment)
+
+      render json: { status: "created" }
+      # ask ben about this elsif
```
ahaha oops 🙃

Zhao-Andy commented on app/controllers/comments_controller.rb:
```diff
@@ -110,6 +110,35 @@ def create
+    elsif (@comment = Comment.where(body_markdown: @comment.body_markdown,
+                                    commentable_id: @comment.commentable.id,
+                                    ancestry: @comment.ancestry)[1])
+      @comment.destroy
```
Re: conversation with Ben, the whole `elsif` clause is to prevent duplicate comments if a user were to hit the submit button twice really quickly or something similar.

rhymes replied:
I think we should add that info in a comment before the elsif, so that I won't ask you again in 5 days :D

Zhao-Andy commented on db/migrate/20200109155115_add_actor_id_to_comments_table.rb:
```diff
@@ -0,0 +1,5 @@
+class AddActorIdToCommentsTable < ActiveRecord::Migration[5.2]
```
This adds a quick way to log who submitted a moderator comment. Used in:
```ruby
# CommentsController#moderator_create:124ish
@comment.actor_id = current_user.id
```

rhymes replied:
What do you think about the name `moderator_id` ?

Zhao-Andy replied:
Yeah personally like `moderator_id` more, but am fairly indifferent too 🙃  Thoughts @benhalpern?

rhymes replied:
Is that actor means so many things in software development and outside :D Moderator seems more self explanatory

Zhao-Andy commented on app/controllers/users_controller.rb:
```diff
@@ -213,6 +213,8 @@ def handle_settings_tab
       handle_pro_membership_tab
     when "account"
       handle_account_tab
+    when "response-templates"
```
Updated everything to response templates 👍

rhymes commented:
@Zhao-Andy since you're working on it I re-added `[WIP]`, feel free to remove it and re-request reviews when you're done

Zhao-Andy commented:
Latest commits should make this work for comments show/index page, and not break anything else along the way. I had to move the logic around in a way that IMO makes things more confusing... Definitely worth a refactor at some point (probably by me) but since this should work and has been a longstanding PR I'd say it's good to go once CI passes.

Also added a "loading screen" but only for the `article#show` comment form for now. The fetch for response templates also only happens once -- for the `article#show` comment form, it will fetch on the clicked the response template button. For the `comment#index/show`, it will fetch (only once) when a reply button is clicked.

Zhao-Andy commented:
@rhymes I've updated the PR with your requests; did not go through with refactoring the JS CodeClimate issues after all. I'd rather leave that for a separate PR. Rest of it should be okay, but looks like I might need to update the test coverage. 😩 

Anyway feel free to review or not dependent on your priorities -- I added you since you've been reviewing this whole time and have the context. :)

rhymes commented:
Hi @Zhao-Andy, let me know if you want to address the first request for change (the internal dashboard is a bit broken) or you'd rather do it in a following PR, don't know how important it is that it's not fully working:

> I couldn't create an email template (neither plain text, nor markdown, nor HTML let me through), both the email and the abuse report email, I think it's because the HTML in /internal/response_templates contains the wrong field names

vaidehijoshi commented:
Ooooh this is basically what we were trying to do when we added `welcome_user_id` into the SiteConfig. Seems like we can just use your `mascot_user_id` instead! 😎

nickytonline commented on app/assets/javascripts/utilities/buildCommentFormHTML.js.erb:
```diff
@@ -19,31 +19,41 @@ function buildCommentFormHTML(commentableId, commentableType, parentId) {
 function cancel(event, el) {
   event.preventDefault();
-  replaceActionButts(el.parentNode.parentNode.parentNode.parentNode);
+  replaceActionButts(el.parentNode.parentNode.parentNode.parentNode.parentNode);
```
What is `el.parentNode.parentNode.parentNode.parentNode.parentNode` referencing? There may be a better way to get a reference to it. If not, consider creating a variable so it can be described and/or consider a comment explaining what it refers to.

nickytonline commented on app/assets/stylesheets/settings.scss:
```diff
@@ -561,7 +579,7 @@
       font-size: 0.9em;
     }
     .github-repo-featured-indicator {
```
I can't find a reference to this CSS selector anywhere in the code except for this file. Are you sure it's used?

nickytonline commented on app/javascript/packs/responseTemplates.js:
```diff
@@ -0,0 +1,35 @@
+import loadingImg from '../../assets/images/loading-ellipsis.svg'
+import { handleLoggedOut, fetchResponseTemplates, addReplyObservers, addToggleListener } from '../responseTemplates/responseTemplates';
+
+const userStatus = document.querySelector('body').getAttribute('data-user-status');
```
```suggestion
const { userStatus } = document.body.dataset;
```

nickytonline commented on app/javascript/packs/responseTemplates.js:
```diff
@@ -0,0 +1,35 @@
+
+  button.addEventListener('click', () => {
+    const responsesWrapper = document.querySelector('.mod-responses-container');
+    responsesWrapper.style.display = 'flex';
```
Can this `display: flex` be added to a CSS class instead? Less JS and in a CSS class it gets cached.

nickytonline commented on app/javascript/packs/responseTemplatesComments.js:
```diff
@@ -0,0 +1,17 @@
+import { handleLoggedOut, addReplyObservers } from '../responseTemplates/responseTemplates';
+
+const userStatus = document.querySelector('body').getAttribute('data-user-status');
```
Same as above

```suggestion
const { userStatus } = document.body.dataset;
```

nickytonline commented on app/javascript/responseTemplates/responseTemplates.js:
```diff
@@ -0,0 +1,225 @@
+/* eslint-disable no-alert */
+/* eslint-disable no-restricted-globals */
+function buildModResponseHTML(response) {
+  const array = response
```
consider a more descriptive name for the variable instead of the generic `array`.

nickytonline commented on app/javascript/responseTemplates/responseTemplates.js:
```diff
@@ -0,0 +1,225 @@
+        ${modResponseHTML}
+        <header><h3>Personal Response Templates</h3></header>
+        ${personalResponseHTML}
+        <a target="_blank" rel="noopener nofollow" href="/settings/response-templates" class="mod-response-create-new">Create new template</a>
```
+1 for `rel="noopener nofollow"`. 👏🏻

nickytonline commented on app/javascript/responseTemplates/responseTemplates.js:
```diff
@@ -0,0 +1,225 @@
+
+  const observer = new MutationObserver(callback);
+
+  targetNodes.forEach(node => {
```
With this, n target nodes get observed which could be quite expensive. It may make more sense to observe the parent node of all the target nodes.

Zhao-Andy commented:
Closing in favor of splitting this into smaller PRs.
